### PR TITLE
Fixed AvoidCatchingNPE and AvoidThrowingRawExceptionTypes issues found by PMD

### PR DESCRIPTION
**Pull request by VelinYordanov**
Fixed pmd issues found is smarthome - AvoidCatchingNPE and AvoidThrowingRawExceptionTypes.

Signed-off-by: VelinYordanov <velin.iordanov@gmail.com>

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java [hunk 1]
```diff
@@ -29,20 +29,28 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * This is an abstract class that can be used when implementing any module handler that handles scripts.
+ * This is an abstract class that can be used when implementing any module
```
Comment (htreu) [inline bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java]: Please check your formatter settings. The maximum line width for ESH is 120.

File: bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java [hunk 2]
```diff
@@ -90,24 +98,27 @@ public void dispose() {
...
         Object type = module.getConfiguration().get(SCRIPT_TYPE);
         Object script = module.getConfiguration().get(SCRIPT);
-        if (type == null || !(type instanceof String) || ((String) type).trim().isEmpty()) {
-            throw new RuntimeException(
+        if (isNotValid(type) || isNotValid(script)) {
+            throw new IllegalStateException(
                     String.format("Type is missing in the configuration of module '%s'.", module.getId()));
-        } else if (script == null || !(script instanceof String) || ((String) script).trim().isEmpty()) {
-            throw new RuntimeException(
-                    String.format("Script is missing in the configuration of module '%s'.", module.getId()));
```
Comment (htreu) [inline bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java]: With this removal the individual error description gets lost.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java [hunk 3]
```diff
@@ -174,7 +174,7 @@ public Void run() {
                 if (cause instanceof RuntimeException) {
                     throw (RuntimeException) cause;
                 } else {
-                    throw new RuntimeException(cause);
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java]: With the lines above this looks like the RuntimeException is thrown on purpose here. @sobimibos can you clarify here what we can do alternatively? Thanks.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java [hunk 4]
```diff
@@ -232,7 +232,7 @@ public HttpClient run() {
             if (cause instanceof RuntimeException) {
                 throw (RuntimeException) cause;
             } else {
-                throw new RuntimeException(cause);
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java]: Same as above.

File: bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java [hunk 5]
```diff
@@ -164,7 +164,7 @@ private Response createResponse(Status status, Object entity) {
             PipedInputStream in = new PipedInputStream(out);
             rp.entity(in);
         } catch (IOException e) {
-            throw new RuntimeException(e);
+            throw new IllegalStateException("I/O error occurred"+e);
```
Comment (htreu) [inline bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java]: please add the original exception as a parameter here.

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java [hunk 6]
```diff
@@ -493,15 +509,17 @@ private void notifyLightStatusListeners(final FullLight fullLight, final String
...
+     * first we check for the common properties which are set for every light type.
+     * If they equal, we additionally try to check the colorMode. If we get an NPE,
+     * the light does not support color mode and the common properties equality is
+     * our result: true. Otherwise if no NPE occurs the equality of colorMode is our
+     * result.
      *
-     * @param state1 Reference state
-     * @param state2 State which is checked for equality.
+     * @param state1
+     *            Reference state
```
Comment (htreu) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeHandler.java]: this line break is also not standard with the ESH code formatter.

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueBridge.java [hunk 7]
```diff
@@ -161,24 +168,25 @@ public boolean isAuthenticated() {
         String lastScan = safeFromJson(result.getBody(), NewLightsResponse.class).lastscan;
 
         switch (lastScan) {
-            case "none":
+        case "none":
```
Comment (htreu) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueBridge.java]: this is also not ESH formatter conform.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java (hunk 3 above)
Comment (VelinYordanov) [reply bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java]: I discarded the changes and added an exception for this class for now.

File: bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java [hunk 8]
```diff
@@ -90,23 +90,25 @@ public void dispose() {
...
+            throw new IllegalStateException(String.format("Type is missing in the configuration of module '%s'.", module.getId()));
+        } else if(isNotValid(script)) {
+            throw new IllegalStateException(String.format("Script is missing in the configuration of module '%s'.", module.getId()));
         } else {
             this.type = (String) type;
             this.script = (String) script;
         }
     }
+    
+    private boolean isNotValid(Object parameter) {
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java]: I see that we would like to check if the parameter is not valid in the cases above.
But I don't think that methods that check that something is not true should be used.
If we want to check if a collection is not empty, we call `!collection.isEmpty()` instead of `collection.isNotEmpty()`.
I would also prefer to provide a method `isValid(Object parameter)` and use `!isValid(type)` above.

File: bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java [hunk 9]
```diff
@@ -90,23 +90,25 @@ public void dispose() {
...
+        return (parameter == null) || !(parameter instanceof String) || ((String) parameter).trim().isEmpty();
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.module.script/src/main/java/org/eclipse/smarthome/automation/module/script/internal/handler/AbstractScriptModuleHandler.java]: Why do we add braces that are not necessary? At least the one around `parameter == null` could be removed.

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueBridge.java [hunk 10]
```diff
@@ -873,6 +873,10 @@ private void handleErrors(Result result) throws IOException, ApiException {
                 }
 
                 for (ErrorResponse error : errors) {
+                    if(error == null) {
```
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueBridge.java]: It seems the formatter is still not correctly applied to all changes.

File: tools/static-code-analysis/pmd/suppressions.properties [hunk 11]
```diff
@@ -3,3 +3,8 @@ org.eclipse.smarthome.io.console.karaf.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.CommandWrapper=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.core.internal.events.ThreadedEventHandler=EmptyIfStmt
+org.eclipse.smarthome.config.xml.ConfigDescriptionConverter=AvoidCatchingNPE
```
Comment (maggu2810) [inline tools/static-code-analysis/pmd/suppressions.properties]: Is it still necessary to catch NPE?
Can you point me to the places where it cannot be checked before?

Comment (VelinYordanov) [reply tools/static-code-analysis/pmd/suppressions.properties]: I think the check should be suppressed for ModelRepositoryImpl and JavaTest. The only reason for the rest is that removing the catching of NPE would lead to longer and a bit uglier code. I could refactor them as well if you prefer.

**Comment by VelinYordanov** [comment]
I also fixed all SimplifyBooleanExpressions issues.

File: tools/static-code-analysis/pmd/suppressions.properties (hunk 11 above)
Comment (maggu2810) [reply tools/static-code-analysis/pmd/suppressions.properties]: Sorry for the delay...

> The only reason for the rest is that removing the catching of NPE would lead to longer and a bit uglier code. I could refactor them as well if you prefer.

I cannot judge without having a look at.
Test for null and prevent NPEs in front of instead of catching it normally result into a cleaner (perhaps this depends on the POV) code. I will have a short look at that classes and try to understand why the code gets "uglier".

File: tools/static-code-analysis/pmd/suppressions.properties [hunk 12]
```diff
@@ -3,3 +3,8 @@ org.eclipse.smarthome.io.console.karaf.internal.OSGiConsole=SystemPrintln
...
+org.eclipse.smarthome.io.rest.core.internal.thing.ThingResource=AvoidCatchingNPE
+org.eclipse.smarthome.test.java.JavaTest=AvoidCatchingNPE
+org.eclipse.smarthome.model.core.internal.ModelRepositoryImpl=AvoidCatchingNPE
```
Comment (maggu2810) [inline tools/static-code-analysis/pmd/suppressions.properties]: WRT https://github.com/eclipse/smarthome/issues/3335 and the Xtext magic, I don't care -- it's black magic for me :wink:

File: tools/static-code-analysis/pmd/suppressions.properties [hunk 13]
```diff
@@ -3,3 +3,8 @@ org.eclipse.smarthome.io.console.karaf.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.CommandWrapper=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.core.internal.events.ThreadedEventHandler=EmptyIfStmt
+org.eclipse.smarthome.config.xml.ConfigDescriptionConverter=AvoidCatchingNPE
+org.eclipse.smarthome.io.rest.core.internal.thing.ThingResource=AvoidCatchingNPE
+org.eclipse.smarthome.test.java.JavaTest=AvoidCatchingNPE
```
Comment (maggu2810) [inline tools/static-code-analysis/pmd/suppressions.properties]: Perhaps okay for relaxed test writings, but hey, we are using pure Java tests and no Groovy, so should be easy to fix (you could also do it in a separate PR and create an tracker issue now if too much tests needs to be fixed).

File: tools/static-code-analysis/pmd/suppressions.properties [hunk 14]
```diff
@@ -3,3 +3,8 @@ org.eclipse.smarthome.io.console.karaf.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.CommandWrapper=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.core.internal.events.ThreadedEventHandler=EmptyIfStmt
+org.eclipse.smarthome.config.xml.ConfigDescriptionConverter=AvoidCatchingNPE
+org.eclipse.smarthome.io.rest.core.internal.thing.ThingResource=AvoidCatchingNPE
```
Comment (maggu2810) [inline tools/static-code-analysis/pmd/suppressions.properties]: Shouldn't this be easy to fix? At which places do we expect that an NPE is thrown?

File: tools/static-code-analysis/pmd/suppressions.properties (hunk 11 above)
Comment (maggu2810) [inline tools/static-code-analysis/pmd/suppressions.properties]: This one is rather easy to fix.
The URI constructor will throw a NPE if the provided text is null only (AFAIK).

Comment (maggu2810) [reply tools/static-code-analysis/pmd/suppressions.properties]: Shouldn't this already do the job?

```diff
@@ -62,12 +62,15 @@ public class ConfigDescriptionConverter extends GenericUnmarshaller<ConfigDescri
             // the URI could be overridden by a context field if it could be
             // automatically extracted
             uriText = (String) context.get("config-description.uri");
+            if (uriText == null) {
+                throw new ConversionException("Missing URI.");
+            }
         }
 
         URI uri = null;
         try {
             uri = new URI(uriText);
-        } catch (NullPointerException | URISyntaxException ex) {
+        } catch (URISyntaxException ex) {
             throw new ConversionException(
                     "The URI '" + uriText + "' in node '" + reader.getNodeName() + "' is invalid!", ex);
         }
```

Comment (VelinYordanov) [reply tools/static-code-analysis/pmd/suppressions.properties]: Yes, I updated the PR with the required changes.

File: tools/static-code-analysis/pmd/suppressions.properties [hunk 15]
```diff
@@ -3,3 +3,5 @@ org.eclipse.smarthome.io.console.karaf.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.CommandWrapper=SystemPrintln
 org.eclipse.smarthome.io.console.rfc147.internal.OSGiConsole=SystemPrintln
 org.eclipse.smarthome.core.internal.events.ThreadedEventHandler=EmptyIfStmt
+org.eclipse.smarthome.model.core.internal.ModelRepositoryImpl=AvoidCatchingNPE
+org.eclipse.smarthome.io.net.http.internal.SecureHttpClientFactory=AvoidThrowingRawExceptionTypes
```
Comment (maggu2810) [inline tools/static-code-analysis/pmd/suppressions.properties]: One question remaining about that one (so, this is perhaps more addresses to the author of that file and other ones caring about the API):
* Why not using a more specific exception instead of encapsulate all ones into `RuntimeException` (at the relevant places)?
* Why not using a checked exception instead of an unchecked? If there is a potential error the caller should handle it correctly. Now, the caller gets not informed (without reading the called function implementation) that he needs to handle that error.

**Comment by VelinYordanov** [comment]
I added a suppression for the `SecureHttpClientFactory` in regards to @htreu 's review

> With the lines above this looks like the RuntimeException is thrown on purpose here. @sobimibos can you clarify here what we can do alternatively? Thanks.

**Comment by maggu2810** [comment]
> * Why not using a more specific exception instead of encapsulate all ones into `RuntimeException` (at the relevant places)?
> * Why not using a checked exception instead of an unchecked? If there is a potential error the caller should handle it correctly. Now, the caller gets not informed (without reading the called function implementation) that he needs to handle that error.

 @sobimibos can you please give us some details?

**Comment by maggu2810** [comment]
@VelinYordanov I added another issue, so this one could be merged as soon as Travis CI succeeded.
